### PR TITLE
[SPARK-10048][SPARKR] Support arbitrary nested Java array in serde.

### PR DESCRIPTION
**Pull request by sun-rui**
This PR:
1. supports transferring arbitrary nested array from JVM to R side in SerDe;
2. based on 1, collect() implemenation is improved. Now it can support collecting data of complex types
   from a DataFrame.

### CODE REVIEW

**Comment by shivaram** [comment]
Thanks @sun-rui I'll take a look at this today

cc @davies

File: core/src/main/scala/org/apache/spark/api/r/SerDe.scala [hunk 1]
```diff
@@ -210,22 +213,31 @@ private[spark] object SerDe {
       writeType(dos, "void")
     } else {
       value.getClass.getName match {
+        case "java.lang.Character" =>
```
Comment (davies) [inline core/src/main/scala/org/apache/spark/api/r/SerDe.scala]: Is this needed?

**Comment by davies** [comment]
@sun-rui Generally, the changes looks good to me, could you add unit tests for ArrayType? Do we want to support create create DataFrame from ArrayType (could be another PR).

**Comment by sun-rui** [comment]
@davies , that will be done in another PR.

File: core/src/main/scala/org/apache/spark/api/r/SerDe.scala (hunk 1 above)
Comment (sun-rui) [inline core/src/main/scala/org/apache/spark/api/r/SerDe.scala]: Not sure. Just for completeness of handling of primitive types.

**Comment by shivaram** [comment]
@sun-rui We can do the `createDataFrame` in another PR, but for this PR can we add a test case using a JSON file which has an `ArrayType` in it ?

**Comment by shivaram** [comment]
BTW does this fix SPARK-9302 as well or does that require `struct` support ?

**Comment by sun-rui** [comment]
@shivaram, now ArrayType in a DataFrame is still not supported, as ArrayType's class is something like scala.collection.mutable.WrappedArray$ofRef, it will be passed as a jobj to R side. So some conversion needs to be done to covert it to Java Array, so that SerDe can pass its content to R side. I planned to do it in another PR as https://issues.apache.org/jira/browse/SPARK-10049, do you want me to do it in this PR?

As for SPARK-9302, it needs support for both ArrayType and StructType. Once these two types are supported, it should be fixed.

**Comment by shivaram** [comment]
I see. I think a separate PR for ArrayType is fine, its just harder to review this change if we can't test it. Its ok, I will take one more closer look tomorrow

**Comment by sun-rui** [comment]
@shivaram, I tried to support ArrayType. By adding code like:
      // Convert Seq[Any] to Array[Any]
      val value =
        if (obj.isInstanceOf[Seq[Any]]) {
          obj.asInstanceOf[Seq[Any]].toArray
        } else {
          obj
        }
I can collect successfully ArrayType on R side.
However, this confilicts with listToSeq() R util functions, because listToSeq() expects to get a jobj to the seq, while SerDe passes back the content of seq instead of jobj. So I think we need some mechanism telling the RBackend what we want a jobj or its content?

Adding support for ArrayType would make this PR big and hard to review. I would be better you could look at this PR and merge it. 

I know you have a concern as we don't have test cases for this PR. Maybe I can add test cases for SerDe:
  Add a function called Echo() in RBackend, it is for test-only purpose. It can pass args back to R side. args could be array, nested array, etc...
What do you think?

**Comment by shivaram** [comment]
Its fine. Lets not complicate this PR with the `listToSeq` thing as that has its own issues like you mention.
I think just adding unit tests to SerDe is a good idea in general and we can either have this `Echo` function or just try to use `identity` in Scala if that works.

File: R/pkg/R/DataFrame.R [hunk 2]
```diff
@@ -628,18 +628,49 @@ setMethod("dim",
...
+              # listCols is a list of columns
+              listCols <- callJStatic("org.apache.spark.sql.api.r.SQLUtils", "dfToCols", x@sdf)
+              stopifnot(length(listCols) == ncol)
+              
+              # An empty data.frame with 0 columns and number of rows as collected
+              nrow <- length(listCols[[1]])
+              if (nrow <= 0) {
+                df <- data.frame()
+              } else {
+                df <- data.frame(row.names = c(1 : nrow))
```
Comment (shivaram) [inline R/pkg/R/DataFrame.R]: I guess `1:nrow` is enough here ? (no need for `c()`)

File: R/pkg/R/DataFrame.R [hunk 3]
```diff
@@ -628,18 +628,49 @@ setMethod("dim",
...
+              
+              # Append columns one by one
+              for (colIndex in 1 : ncol) {
+                # Note: appending a column of list type into a data.frame so that
+                # data of complex type can be held. But getting a cell from a column
+                # of list type returns a list instead of a vector. So for columns of
+                # non-complex type, append them as vector.
+                col <- listCols[[colIndex]]
+                if (length(col) <= 0) {
+                  df[[names[colIndex]]] <- col
```
Comment (shivaram) [inline R/pkg/R/DataFrame.R]: this is just to handle empty columns ? Can't we just return from line 644 if one of the columns is empty ? Or do we want to handle cases where some columns are empty and some are not etc.

File: sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala [hunk 4]
```diff
@@ -98,27 +98,20 @@ private[r] object SQLUtils {
     val bos = new ByteArrayOutputStream()
     val dos = new DataOutputStream(bos)
 
-    SerDe.writeInt(dos, row.length)
-    (0 until row.length).map { idx =>
-      val obj: Object = row(idx).asInstanceOf[Object]
-      SerDe.writeObject(dos, obj)
-    }
+    val cols =
+      (0 until row.length).map { idx =>
```
Comment (shivaram) [inline sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala]: This can be moved to the previous line

**Comment by shivaram** [comment]
I took a more detailed look at the code and I only had some minor comments inline. So I think it looks pretty good but I think taking this opportunity to add some tests to SerDe is a good idea.

**Comment by sun-rui** [comment]
will add test cases.

File: R/pkg/R/DataFrame.R (hunk 2 above)
Comment (sun-rui) [inline R/pkg/R/DataFrame.R]: you are correct

File: R/pkg/R/DataFrame.R (hunk 3 above)
Comment (sun-rui) [inline R/pkg/R/DataFrame.R]: This is special handling for empty columns where the DataFrame has >=1 column but nrow = 0.  If a column is empty (list()), do.call(c, list()) will return NULL. 

Line 644 is for the case where DataFrame with 0 column and 0 row.

File: sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala (hunk 4 above)
Comment (sun-rui) [inline sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala]: OK

File: sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala [hunk 5]
```diff
@@ -98,27 +98,17 @@ private[r] object SQLUtils {
     val bos = new ByteArrayOutputStream()
     val dos = new DataOutputStream(bos)
 
-    SerDe.writeInt(dos, row.length)
-    (0 until row.length).map { idx =>
-      val obj: Object = row(idx).asInstanceOf[Object]
-      SerDe.writeObject(dos, obj)
-    }
+    val cols = (0 until row.length).map { idx => row(idx).asInstanceOf[Object]}.toArray
```
Comment (shivaram) [inline sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala]: we should have one extra space after `[Object]` before `}`. In fact if the whole map fits on one line you can just use `map(idx => ...).toArray`

**Comment by sun-rui** [comment]
@shivaram , test cases for SerDe added. Now the SerDe does not support transferring a list of different element types from R side to JVM side. Let's leave it for future's PR.

File: sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala (hunk 5 above)
Comment (sun-rui) [inline sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala]: fixed.

**Comment by sun-rui** [comment]
rebased to master

**Comment by shivaram** [comment]
Jenkins, retest this please

**Comment by shivaram** [comment]
Jenkins, retest this please

**Comment by shivaram** [comment]
@sun-rui looks like the test failures are related to this PR

**Comment by sun-rui** [comment]
The test passed on my machine, I don't know the reason. Anyway, add spark context initialization into test_Serde to see if it can pass on Jenkins.

File: R/pkg/inst/tests/test_Serde.R [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  y <- callJStatic("SparkRHandler", "echo", x)
+  expect_equal(x, y)
+  expect_equal(class(y[[1]]), "character")
+})
+
+test_that("SerDe of list of lists", {
+  x <- list(list(1L, 2L, 3L), list(1, 2, 3),
+            list(TRUE, FALSE), list("a", "b", "c"))
+  y <- callJStatic("SparkRHandler", "echo", x)
+  expect_equal(x, y)
```
Comment (shivaram) [inline R/pkg/inst/tests/test_Serde.R]: Could we also add some tests with empty columns / empty lists (as we have some code paths just to handle these)

Comment (sun-rui) [inline R/pkg/inst/tests/test_Serde.R]: added.

**Comment by shivaram** [comment]
Jenkins, retest this please

**Comment by shivaram** [comment]
Thanks @sun-rui -- Change LGTM. 

@davies Any other comments ?

**Comment by davies** [comment]
LGTM

**Comment by shivaram** [comment]
Alright I'm merging this to master. Note that I'm not porting this branch-1.5 as I think this is a relatively big change and it doesn't have any immediate feature benefits.

**Pull request closed without merging**